pkg/api/controller.go

```diff
@@ -554,6 +554,40 @@ func (c *Controller) Login(w http.ResponseWriter, r *http.Request, body apigen.L
 	writeResponse(w, r, http.StatusOK, response)
 }
 
+func (c *Controller) ExternalPrincipalLogin(w http.ResponseWriter, r *http.Request, body apigen.ExternalPrincipalLoginJSONRequestBody) {
+	ctx := r.Context()
+	if c.isExternalPrincipalNotSupported(ctx) {
+		writeError(w, r, http.StatusNotImplemented, "Not implemented")
+		return
+	}
+
+	c.LogAction(ctx, "external_principal_login", r, "", "", "")
+
+	userID, err := c.Auth.ExternalPrincipalLogin(ctx, body)
+	if c.handleAPIError(ctx, w, r, err) {
+		if errors.Is(err, ErrAuthenticatingRequest) {
+			writeResponse(w, r, http.StatusUnauthorized, http.StatusText(http.StatusUnauthorized))
+			return
+		}
```

Comment:
- handleAPIError already writes the response, so you just need to return
- It looks like the token generation happens only in case of error (brackets are in the wrong place I think)